feat(acl-v2): ACL v2/RBAC documentation (#9754)
vatesfr/xen-orchestra: docs/docs/acl-v2.md

# ACL v2 (REST API/XO6)

ACL v2 is the access control system for the Xen Orchestra REST API and the XO6 UI. It lets you define exactly what each user or group can see and do — down to individual objects — without granting them full administrator access.

## What changed from v1

The old ACL system (v1) allowed granting access to individual objects (a VM, an SR…). Simple, but limited: there was no way to say _"this user can shutdown only VMs tagged `qa`"_. It also only covered **XAPI objects** — VMs, hosts, SRs, networks. Users, groups, backups, schedules, and jobs were out of scope.

ACL v2 introduces a full **RBAC (Role-Based Access Control)** model with effects, selectors, and an action hierarchy, covering the entire infrastructure including XO management objects.

:::note
ACL v2 is available through the **REST API only**. The JSON-RPC API (used by XO5) stays on ACL v1. Conversely, ACL v1 is not available on the REST API.

> :::

Reviewer: Should we mention somewhere here that the name of the feature might change?

The system is built around two building blocks:

| Concept       | Description                                                                                                             |
| ------------- | ----------------------------------------------------------------------------------------------------------------------- |
| **Role**      | A named container for privileges.                                                                                       |
| **Privilege** | A rule inside a role: which action is allowed (or denied) on which resource, and optionally on which subset of objects. |

### Roles

A role is just a label with a collection of privileges. You create a role once, add privileges to it, then assign it to as many users or groups as needed.

A user's effective privileges are the union of all privileges from all their direct roles and the roles of their groups.

### Privileges

A privilege defines:

- **resource** — the type of object (e.g. `vm`, `backup-job`, `sr`)
- **action** — what operation is allowed or denied (e.g. `read`, `start`, `delete`)

> - **effect** — `allow` or `deny`

Reviewer: ```suggestion
- **effect** — whether the privilege grants (`allow`) or blocks (`deny`) the action
```

> - **selector** _(optional)_ — a filter expression to restrict the privilege to a subset of objects (complex-matcher format)

Reviewer: ```suggestion
- **selector** _(optional)_ — a filter expression to restrict the privilege to a subset of objects ([complex-matcher format](/manage_infrastructure#filter-syntax))
```

| `GET /rest/v0/users/me`                        | Get your own user profile                   |
| `GET /rest/v0/users/me/privileges`             | List your own privileges                    |
| `GET /rest/v0/users/me/authentication_tokens`  | List your own authentication tokens         |
| `POST /rest/v0/users/me/authentication_tokens` | Create an authentication token for yourself |

`me` is a convenience alias — it is automatically redirected to `/rest/v0/users/{your-id}`.

:::tip
The Swagger UI available at `/rest/v0/swagger` documents every endpoint with its required privileges. Endpoints with no declared privilege are admin-only.
:::

---

## Supported resources and actions

### Infrastructure resources

| Resource        | Available actions                                                                                                                                                                    |
| --------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |

> | `vm`            | `read`, `delete`, `export`, `pause`, `start`, `reboot` (`clean`, `hard`), `shutdown` (`clean`, `hard`), `resume`, `snapshot`, `suspend`, `unpause`, `update` (`datasources`, `tags`) |

Reviewer: How does the `` `foo` (`bar`, `baz`) `` syntax work? Would it be simpler to flatten them since there aren't that many of them?

If not, I think we should at least have an example using those actions immediately after the tables to make sure users will understand how to use them.

Author: `reboot (clean, hard)` -> `reboot:clean`, `reboot:hard`.
We will have a lot of these actions in the future (E.g. with update:name_label, update:tags, update:...)

| `vm-controller` | `read`, `update:tags`                                                                                                                                                                |
| `vdi`           | `read`, `create`, `delete`, `boot`, `export-content`, `import-content`, `update:tags`                                                                                                |
| `vdi-snapshot`  | `read`                                                                                                                                                                               |
| `vdi-unmanaged` | `read`                                                                                                                                                                               |
| `vif`           | `read`, `create`                                                                                                                                                                     |

> | `vbd`           | `read`                                                                                                                                                                               |

Reviewer: Do we have a way to make sure we don't forget to update this table as we add more actions?

Author: Not for now.
BTW, the swagger documentation will always be up to date

```json
{ "resource": "vm", "action": "*",    "effect": "allow" }
{ "resource": "vm", "action": "*",    "effect": "deny",  "selector": "tags:prod" }
```

Carol can start, stop, delete, snapshot any VM she wants. Production VMs are completely invisible to her — they never appear in her lists and any direct attempt returns a 403.

---

## Walkthrough: creating and assigning a role

### Step 1 — Create a role

![acl-role create documented in Swagger](./assets/swagger-create-role.png)

Response:

```json
{ "id": "e4b1f3c2-1234-5678-abcd-000000000001" }
```

### Step 2 — Add privileges to the role

![acl-privilege create documented in Swagger](./assets/swagger-create-privilege.png)

### Step 3 — Assign the role to the user

![add user to role documented in Swagger](./assets/swagger-add-user-to-role.png)

The user can now list VMs.

To assign to a group instead (all group members inherit the role):

![add group to role documented in Swagger](./assets/swagger-add-group-to-role.png)

---

## Selectors

> By default, a privilege applies to **all** objects of the given resource type. The optional `selector` field narrows it down using the [complex-matcher](https://docs.xen-orchestra.com/manage_infrastructure#filter-syntax) syntax — the same filter syntax used in the XO UI.

Reviewer: ```suggestion
By default, a privilege applies to **all** objects of the given resource type. The optional `selector` field narrows it down using the [complex-matcher](./manage_infrastructure#filter-syntax) syntax — the same filter syntax used in the XO UI.
```

A selector is evaluated against each object's properties. If it matches, the privilege applies; otherwise it does not.

### Common selector patterns

| Goal                           | Selector example              |
| ------------------------------ | ----------------------------- |
| A specific object by its ID    | `id: <uuid>`                  |
| All objects in a pool          | `$pool: <pool-uuid>`          |
| VMs tagged with a label        | `tags: <tag>`                 |
| VMs by power state             | `power_state: Running`        |
| VMs created by a specific user | `creation:creator: <user-id>` |

---

## Tips

- **Start minimal.** Add only the privileges a user actually needs. It is easy to add more later.
- **Use groups.** Assigning a role to a group avoids repeating the same assignment for every user.

> - **Combine `allow` and `deny`.** Grant broad access with a wildcard privilege, then carve out exceptions with `deny` + a selector.

Reviewer: This kind of goes against the first tip, so I'd mention to only use this when the first tip doesn't apply:

```suggestion
- **Combine `allow` and `deny`.** When a user needs broad access with specific exceptions, grant with a wildcard privilege and carve out exceptions with `deny` + a selector.
```
